src/dataset.cpp

```diff
@@ -339,79 +339,124 @@ void multiply(const gsl::index size, ptr<double> RESTRICT v1,
-              (var1.dimensions() == error2.dimensions())) {
-            // Optimization if all dimensions match, avoiding allocation of
-            // temporaries and redundant streaming from memory of large array.
-            error1.setUnit(var2.unit() * var2.unit() * error1.unit() +
-                           var1.unit() * var1.unit() * error2.unit());
-            var1.setUnit(var1.unit() * var2.unit());
-
-            // TODO We are working with VariableSlice here, so get<> returns a
-            // view, not a span, i.e., it is less efficient. May need to do this
-            // differently for optimal performance.
-            auto v1 = var1.template span<double>();
-            const auto v2 = var2.template span<double>();
-            auto e1 = error1.template span<double>();
-            const auto e2 = error2.template span<double>();
-            // TODO Need to ensure that data is contiguous!
-            aligned::multiply(v1.size(), v1.data(), e1.data(), v2.data(),
-                              e2.data());
-          } else {
-            // TODO Do we need to write this differently if the two operands are
-            // the same? For example,
-            // error1 = error1 * (var2 * var2) + var1 * var1 * error2;
-            error1 *= (var2 * var2);
-            error1 += var1 * var1 * error2;
-            // TODO: Catch errors from unit propagation here and give a better
-            // error message.
-            var1 *= var2;
+  } else if (rhs_var.tag() == Data::Variance) {
+    // Do nothing, math for variance is done when processing corresponding
+    // value.
+  } else {
+    lhs_var *= rhs_var;
+  }
+}
+
+} // namespace
+template <class T1, class T2> T1 &times_equals(T1 &dataset, const T2 &other) {
+  std::set<std::string> lhs_names;
+  for (const auto &lhs_var : other)
+    if (lhs_var.isData())
+      lhs_names.insert(lhs_var.name());
+
+  // See operator+= for additional comments.
+  for (const auto &rhs_var : other) {
+    // Look for exact match in lhs for var in rhs
+    if (dataset.contains(rhs_var.tag(), rhs_var.name())) {
+      auto lhs_var = dataset(rhs_var.tag(), rhs_var.name());
+      if (lhs_var.isCoord()) {
+        // Coordinate variables must match
+        dataset::expect::variablesMatch(lhs_var, rhs_var);
+      } else if (lhs_var.isData()) {
+        // Use slices to capture related variables for example variance data
+        // variables
+        auto lhs_slice = dataset.subset(lhs_var.name());
+        auto rhs_slice = other.subset(rhs_var.name());
+        multiply_slices(lhs_slice, rhs_slice, lhs_var, rhs_var);
+      }
+    } else {
+      // Note that this is handled via name, i.e., there may be values and
+      // variances, i.e., two variables.
+      if (rhs_var.isData() && lhs_names.size() == 1) {
+        // Only a single (named) variable in RHS, operate on all.
+        // Not a coordinate, apply from all.
+        // op([a, b], [a]) = [op(a, a), op(b, a)] is legal
```

Comment:
Typo? As it is, it shouldn't be. `op([a, b], [c]) = [op(a, c), op(b, c)]` would be legal.